common/rfb/CSecurityRSAAES.cxx

```diff
@@ -0,0 +1,444 @@
+  sha1_update(&ctx, serverKey.size, serverKeyE);
+  sha1_digest(&ctx, sizeof(f), f);
+  const char *title = "Server key fingerprint";
+  CharArray text;
+  text.format(
+    "The server has provided the following identifying information:\n"
+    "Fingerprint: %02x-%02x-%02x-%02x-%02x-%02x-%02x-%02x\n"
+    "Please verify that the information is correct and press \"Yes\". "
+    "Otherwise press \"No\"", f[0], f[1], f[2], f[3], f[4], f[5], f[6], f[7]);
+  if (!CSecurityTLS::msg->showMsgBox(UserMsgBox::M_YESNO, title, text.buf))
+    throw AuthFailureException("server key mismatch");
+}
+
+void CSecurityRSAAES::writeRandom()
+{
+  if (gnutls_rnd(GNUTLS_RND_KEY, clientRandom, keySize) < 0)
+    throw AuthFailureException("failed to generate random");
+  mpz_t x;
+  mpz_init(x);
+  if (!rsa_encrypt(&serverKey, NULL, random_func, keySize, clientRandom, x)) {
+    mpz_clear(x);
+    throw AuthFailureException("failed to encrypt random");
+  }
+  rdr::U8* buffer = new rdr::U8[serverKey.size];
+  nettle_mpz_get_str_256(serverKey.size, buffer, x);
+  mpz_clear(x);
+  rawos->writeU16(serverKey.size);
+  rawos->writeBytes(buffer, serverKey.size);
+  rawos->flush();
+  delete[] buffer;
+}
+
+bool CSecurityRSAAES::readRandom()
+{
+  if (!rawis->hasData(2))
+    return false;
+  rawis->setRestorePoint();
+  size_t size = rawis->readU16();
+  if (size != clientKey.size)
+    throw AuthFailureException("client key length doesn't match");
+  if (!rawis->hasDataOrRestore(size))
+    return false;
+  rawis->clearRestorePoint();
+  rdr::U8* buffer = new rdr::U8[size];
+  rawis->readBytes(buffer, size);
+  size_t randomSize = keySize;
+  mpz_t x;
+  nettle_mpz_init_set_str_256_u(x, size, buffer);
+  delete[] buffer;
+  if (!rsa_decrypt(&clientKey, &randomSize, serverRandom, x) ||
+    randomSize != (size_t)keySize) {
+    mpz_clear(x);
+    throw AuthFailureException("failed to decrypt server random");
+  }
+  mpz_clear(x);
+  return true;
+}
+
+void CSecurityRSAAES::setCipher()
+{
+  rdr::U8 key[32];
+  if (keySize == 16) {
+    struct sha1_ctx ctx;
+    sha1_init(&ctx);
+    sha1_update(&ctx, 16, clientRandom);
+    sha1_update(&ctx, 16, serverRandom);
+    sha1_digest(&ctx, 16, key);
+    rais = new rdr::AESInStream(rawis, key, 128);
+    sha1_init(&ctx);
+    sha1_update(&ctx, 16, serverRandom);
+    sha1_update(&ctx, 16, clientRandom);
+    sha1_digest(&ctx, 16, key);
+    raos = new rdr::AESOutStream(rawos, key, 128);
+  } else {
+    struct sha256_ctx ctx;
+    sha256_init(&ctx);
+    sha256_update(&ctx, 32, clientRandom);
+    sha256_update(&ctx, 32, serverRandom);
+    sha256_digest(&ctx, 32, key);
+    rais = new rdr::AESInStream(rawis, key, 256);
+    sha256_init(&ctx);
+    sha256_update(&ctx, 32, serverRandom);
+    sha256_update(&ctx, 32, clientRandom);
+    sha256_digest(&ctx, 32, key);
+    raos = new rdr::AESOutStream(rawos, key, 256);
+  }
+  if (secType == secTypeRA2 || secType == secTypeRA256)
+    cc->setStreams(rais, raos);
```

Comment:
I think it might be clearer to always switch streams, as these will always be active for the rest of the handshake. So instead there will be an `if` if we should switch back or not after the handshake is done.